bundles/mapping/mapwfs2/plugin/components/convertRenderFeatures.js

```diff
@@ -0,0 +1,104 @@
+import LineString from 'ol/geom/LineString';
+import MultiLineString from 'ol/geom/MultiLineString';
+import MultiPoint from 'ol/geom/MultiPoint';
+import MultiPolygon from 'ol/geom/MultiPolygon';
+import Point from 'ol/geom/Point';
+import Polygon from 'ol/geom/Polygon';
+import GeometryLayout from 'ol/geom/GeometryLayout';
+import Feature from 'ol/Feature';
+import RenderFeature from 'ol/render/Feature';
+import {equivalent, getTransform} from 'ol/proj';
+import Units from 'ol/proj/Units';
+import {create as createTransform, compose as composeTransform} from 'ol/transform';
+import {transform2D} from 'ol/geom/flat/transform';
+import {getHeight} from 'ol/extent';
+
+const tmpTransform = createTransform();
+
+/**
+ * @function convertRenderFeatures
+ * Creates Feature instances from RenderFeature instances. Uses OL internal APIs.
+ * @param {ol/render/Feature~RenderFeature[]} inputFeatures
+ * @param {ol/VectorTile} tile
+ * @param {ol/source/VectorTile} source
+ * @return {ol/Feature[]}
+ */
+export default function convertRenderFeatures (inputFeatures, tile, source) {
+    const viewProjection = source.getProjection();
+    if (!inputFeatures.length) {
+        return [];
+    }
+    if (!(inputFeatures[0] instanceof RenderFeature)) {
+        return inputFeatures.slice();
+    }
+
+    const tileProjection = tile.getProjection();
+    let transformer = null;
+    if (!equivalent(tileProjection, viewProjection)) {
+        if (tileProjection.getUnits() === Units.TILE_PIXELS) {
+            transformer = function (coords) {
+                const pixelExtent = tile.getExtent();
+                const projectedExtent = source.getTileGrid().getTileCoordExtent(tile.getTileCoord());
+                const scale = getHeight(projectedExtent) / getHeight(pixelExtent);
+                composeTransform(tmpTransform,
+                    projectedExtent[0], projectedExtent[3],
+                    scale, -scale, 0,
+                    0, 0);
+                transform2D(coords, 0, coords.length, 2,
+                    tmpTransform, coords);
+                return coords;
+            };
+        } else {
+            transformer = getTransform(tileProjection, viewProjection);
+        }
+    }
+
+    return inputFeatures.map((renderFeature) => {
+        const id = renderFeature.getId();
+        const values = renderFeature.getProperties();
+        const ends = renderFeature.getEnds();
+        const geometryType = renderFeature.getType();
+
+        let flatCoordinates = renderFeature.getOrientedFlatCoordinates();
+        if (transformer) {
+            flatCoordinates = transformer(flatCoordinates.slice());
+        }
+        /**
+         *  Geometry conversion algorithm adapdted from
+         *  https://github.com/openlayers/openlayers/blob/e426af29c20c7b08ee323b2ee0f78d34b61c4c34/src/ol/format/MVT.js#L186-L211
+         * */
+        let geom;
+        if (geometryType === GeometryType.POLYGON) {
+            const endss = [];
```

Comment:
More descriptive names that could be used instead of "end", "ends" and "endss"? Or is the point that this is copy/paste from OL and it's easier to track changes if we don't modify it?

Reply:
Yes, the point is that if the variable names are the same as in the OL source, it's easier to see that the code is the same.